Made RuleSetMappingDefinition immutable

usarskyy opened the pull request:
Query returns are initialized immediately (as an array) in order to get rid of unnecessary List->Array conversions

hazzik commented on src/NHibernate/Cfg/XmlHbmBinding/ResultSetMappingBinder.cs:
```diff
@@ -29,19 +31,9 @@ public ResultSetMappingDefinition Create(HbmSqlQuery sqlQuerySchema)
-			}
-
-			return definition;
+			var returns = items?.Select(CreateQueryReturn).Where(x => x != null) ?? Enumerable.Empty<INativeSQLQueryReturn>();
```
No C# 6 support yet

hazzik commented on src/NHibernate/Engine/ResultSetMappingDefinition.cs:
```diff
@@ -2,36 +2,30 @@
+
+		public string Name => name;
+
+		public INativeSQLQueryReturn[] QueryReturns => queryReturns;
```
No C# 6 support yet.

usarskyy commented:
Updated

oskarb commented:
It fails because it still uses "nameof" operator. On the other hand this changes a public API so must be punted to NH 5.0, where perhaps we should allow C# 6.

The null check on the return value from CreateQueryReturn() got lost.

It's a misnomer to claim it's immutable, because it now provides access to the array, and the array is certainly not immutable. So in a manner of speaking, ResultSetMappingDefinition is not less immutable than it was before. That array in turn gets pushed around a lot which I'm not too fond of. Perhaps it should instead return IReadOnlyList<T> or something? Perhaps the places that accept an INativeSQLQueryReturn[] should instead accept an IReadOnlyList<T>? Or will that have a noticeable performance impact on the various loops over queryReturns?

usarskyy commented:
> It fails because it still uses "nameof" operator.

Sorry, it had to be removed but I missed it.

> On the other hand this changes a public API so must be punted to NH 5.0, where perhaps we should allow C# 6.

I didn't know when my pull request will be accepted, so I made some non-backward compatible API changes.
I checked how ResultSetMappingDefinition is used and it looks like making it completely immutable is better than current implementation. 
If somebody wants to change ResultSetMappingDefinition instance then I would say that creating new ResultSetMappingDefinition instance is safer approach rather than updating internal collections of the existing object.

> The null check on the return value from CreateQueryReturn() got lost.

Actually it is still there but expressed as ".Where(x => x != null)":

```

var returns = itemsNullSafe.Select(CreateQueryReturn)
 +																 .Where(x => x != null)
 +																 .ToArray();
```

> It's a misnomer to claim it's immutable, because it now provides access to the array, and the array is certainly not immutable

I referred to the internal collection itself. Previously list could be changed even that it was declared as readonly. I wanted to have it always fixed. But array items are for sure not immutable.

> That array in turn gets pushed around a lot which I'm not too fond of. Perhaps it should instead return IReadOnlyList or something?

Yes, you are right. IReadOnlyList makes more sense. But if I remember it correct, this change required more non-backward compatible API changes.
If you are OK with IReadOnlyList-based implementation, I could replace array with IReadOnlyList.

> Or will that have a noticeable performance impact on the various loops over queryReturns?

I don't know how other projects use ResultSetMappingDefinition but looking at NH sources I don't see that there will be any noticeable performance impact.

hazzik commented:
Hey, if you rebase on top of master, the C#6 shall be enabled there and
working
On Mon, 1 Aug 2016 at 7:16 AM, Volodymyr Usarskyy <notifications@github.com>
wrote:

> It fails because it still uses "nameof" operator.
>
> Sorry, it had to be removed but I missed it.
>
> On the other hand this changes a public API so must be punted to NH 5.0,
> where perhaps we should allow C# 6.
>
> I didn't know when my pull request will be accepted, so I made some
> non-backward compatible API changes.
> I checked how ResultSetMappingDefinition is used and it looks like making
> it completely immutable is better than current implementation.
> If somebody wants to change ResultSetMappingDefinition instance then I
> would say that creating new ResultSetMappingDefinition instance is safer
> approach rather than updating internal collections of the existing object.
>
> The null check on the return value from CreateQueryReturn() got lost.
>
> Actually it is still there but expressed as ".Where(x => x != null)":
>
>
> var returns = itemsNullSafe.Select(CreateQueryReturn)
>  +                                                               .Where(x => x != null)
>  +                                                               .ToArray();
>
> It's a misnomer to claim it's immutable, because it now provides access to
> the array, and the array is certainly not immutable
>
> I referred to the internal collection itself. Previously list could be
> changed even that it was declared as readonly. I wanted to have it always
> fixed. But array items are for sure not immutable.
>
> That array in turn gets pushed around a lot which I'm not too fond of.
> Perhaps it should instead return IReadOnlyList or something?
>
> Yes, you are right. IReadOnlyList makes more sense. But if I remember it
> correct, this change required more non-backward compatible API changes.
> If you are OK with IReadOnlyList-based implementation, I could replace
> array with IReadOnlyList.
>
> Or will that have a noticeable performance impact on the various loops
> over queryReturns?
>
> I don't know how other projects use ResultSetMappingDefinition but looking
> at NH sources I don't see that there will be any noticeable performance
> impact.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/nhibernate/nhibernate-core/pull/475#issuecomment-236450855>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAI1lytzibFAXooiS1EWRmf-e6So2JaQks5qbPRygaJpZM4ItWhB>
> .
>

usarskyy commented:
Hi!

Sorry for the delay with an update. Hard to find time to get back to this issue.

How shall I proceed with new changes? Shall I get latest master with C# 6 enabled, apply changes from current pull request, add IReadOnlyList refactorings and create new pull request?

fabiomaulo commented:
C#6 is just a matter of the compiler.
If is acceptable that everybody have to have at least VS2015 (community or
not), the usage of the new compiler does not mean a big problem for the
actual NH version.

If you have VS2015 and you need to fix the version of the compiler for a
specific project...

[image: Inline image 1]

cu


On Sun, Jul 31, 2016 at 10:16 AM, Oskar Berggren <notifications@github.com>
wrote:

> It fails because it still uses "nameof" operator. On the other hand this
> changes a public API so must be punted to NH 5.0, where perhaps we should
> allow C# 6.
>
> The null check on the return value from CreateQueryReturn() got lost.
>
> It's a misnomer to claim it's immutable, because it now provides access to
> the array, and the array is certainly not immutable. So in a manner of
> speaking, ResultSetMappingDefinition is not less immutable than it was
> before. That array in turn gets pushed around a lot which I'm not too fond
> of. Perhaps it should instead return IReadOnlyList or something? Perhaps
> the places that accept an INativeSQLQueryReturn[] should instead accept an
> IReadOnlyList? Or will that have a noticeable performance impact on the
> various loops over queryReturns?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nhibernate/nhibernate-core/pull/475#issuecomment-236429361>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHob2tM7TC7ipMzY72nGiy8Aq82MkKDks5qbKAlgaJpZM4ItWhB>
> .
>



-- 
--
Fabio Maulo
*T* @fabiomaulo

usarskyy commented:
> C#6 is just a matter of the compiler.

Not only. In this case it is also a matter of new .NET Framework. 
In order to use IReadOnlyList I will need to update a project to .NET 4.5 (I branched it with .NET 4).

But I was mainly asking about new pull request because of the new changes that require more (non-backward compatible) API changes and upgrade to .NET 4.5 (if it was not yet done). This is a bit out of scope of my initial pull request.

usarskyy commented:
@oskarb @hazzik  I rebased on latest master as you suggested but I can't use IReadOnlyList<> because NHibernate project still uses .NET Framework 4. Shall I upgrade it to 4.5?

oskarb commented:
Have we decided to make NH4.1 require .Net 4.5?

fabiomaulo commented:
Instead IReadOnlyList you can use an implementation of ICollection
(wrapping, for example, a List) where:
 IsReadOnly => true
Add => throw NotAllowedException
Clear => throw NotAllowedException
Remove => throw NotAllowedEception

The IReadOnlyList is not mandatory to achive the target required by Oskar.

On Sat, Sep 10, 2016 at 7:31 AM, Oskar Berggren <notifications@github.com>
wrote:

> Have we decided to make NH4.1 require .Net 4.5?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nhibernate/nhibernate-core/pull/475#issuecomment-246104519>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHob_sze-4qB5Nd4TqHRY4KolC7QIrOks5qoob6gaJpZM4ItWhB>
> .
>



-- 
--
Fabio Maulo
*T* @fabiomaulo

usarskyy commented:
Yes, I know that I could implement it like you suggested but the whole point of using IReadOnlyCollection<T> is to show that API returns readonly collection without need to check internal properties. 
Also with IReadOnlyCollection compiler will not let you write things like:

`items[0] = null;`

(which is perfectly valid code even if "IsReadOnly == true" in List<T> instance)

If I am forced to use .NET 4 types then it would rather use ReadOnlyCollection<T> that is already present in .NET 4. In future releases, when NH migrates to .NET 4.5, we can easily change return type to IReadOnlyCollection<T>.

_BTW, with custom implementation of ICollection<T> as you suggested, we loose indexers. This, of course, can be fixed if we implement IList<T> but then we come back to "need to check internal properties"._

fabiomaulo commented:
The ToArray creates a new instance of the array (
http://referencesource.microsoft.com/#System.Core/System/Linq/Enumerable.cs,ed118118b642d9d4)
so, whatever you do with it will not change the state os the field
queryReturns.
From who we are looking to protect that array ? who is the enemy ? ;)

On Sat, Sep 10, 2016 at 9:39 AM, Volodymyr Usarskyy <
notifications@github.com> wrote:

> Yes, I know that I could implement it like you suggested but the whole
> point of using IReadOnlyCollection is to show that API returns readonly
> collection without need to check internal properties.
> Also with IReadOnlyCollection compiler will not let you write things like:
>
> items[0] = null;
>
> (which is perfectly valid code even if "IsReadOnly == true" in List
> instance)
>
> If I am forced to use .NET 4 types then it would rather use
> ReadOnlyCollection that is already present in .NET 4. In future releases,
> when NH migrates to .NET 4.5, we can easily change return type to
> IReadOnlyCollection.
>
> *BTW, with custom implementation of ICollection as you suggested, we loose
> indexers. This, of course, can be fixed if we implement IList but then we
> come back to "need to check internal properties".*
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nhibernate/nhibernate-core/pull/475#issuecomment-246109609>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHob4nvJGMmXHaWNRyxU5TnFWTPrqObks5qoqUWgaJpZM4ItWhB>
> .
>



-- 
--
Fabio Maulo
*T* @fabiomaulo

usarskyy commented:
The whole current implementation does not make sense. Method "AddQueryReturn" is never called and in fact should never be called. ResultSetMappingDefinition is treated as immutable object BUT its API does not fix this behavior.
So, we keep "queryReturns" in list even that it is not required. The we keep converting list to array which does not make sense because list items are fixed.

> so, whatever you do with it will not change the state os the field queryReturns.
In the original version I can change state of this field. All methods are fully exposed.

> From who we are looking to protect that array ? who is the enemy ? ;)
Whoever have a wrong belief that changing "queryReturns" field might be a good idea.

FYI, my change is not about performance. I just want to fix the API.

hazzik commented:
@oskar. 4.x branch stays .NET 4; 5.x - moves to .NET 4.5./4.6.2

@usarskyy, I meant only compiler and language features of C# 6, not .NET Framework.

usarskyy commented:
@hazzik OK. In this case, shall I use ReadOnlyCollection (part of .NET 4) as return type and push to master? Or, use IReadOnlyList (part of .NET 4.5) and push to 5.0.x branch?

usarskyy commented:
Ok, I updated my pull request based on the assumption that it will be merged with one of the 4.* branches.
I used ReadOnlyCollection<> from .NET 4. It has compatible API with IReadOnlyList<>, so the code can be easily updated for future NH 5 release.

One thing that I am not sure about are changes in SqlQueryImpl. There were a few unused internal constructors which I removed. First I refactored them to use ReadOnlyCollection but then I decided that there is no use of them anyway and dropped that dead code.

fredericDelaporte commented:
If you rebase, master compile now against .Net Framework 4.6.1 with MSBuild 2017 (c#7 available).

hazzik commented:
Targeting 6.0 as the PR has extensive API interface changes

bahusoid commented:
This PR contains some questionable and unrelated changes. Notably I don't like usages of `ImmutableAddRange` and `ImmutableAdd` in `SqlQueryImpl`. That'ts mutable class so why bother?

>in order to get rid of unnecessary List->Array conversions

So one conversion in `RuleSetMappingDefinition ` is removed but N conversions for each `SqlQueryImpl.Add...` call are added instead.

hazzik commented:
Closing as per @bahusoid.
